simplesat/sat/policy/undetermined_clause_policy.py

```diff
@@ -89,13 +93,12 @@ def get_next_package_id(self, assignments, clauses):
         assert assignments.get(candidate_id) is None, \
             "Trying to assign to a variable which is already assigned."
 
-        if not self.prefer_installed:
-            # If this exact package version is available locally, use it.
-            # NOTE: when repository priority is implemented, this will cause
-            # the virtual <installed> repository to act as if it always has the
-            # highest priority.
-            key = self._package_key(candidate_id)
-            candidate_id = self._preferred_package_ids.get(key, candidate_id)
+        # If this exact package version is available locally, use it.
+        # NOTE: when repository priority is implemented, this will cause
+        # the virtual <installed> repository to act as if it always has the
+        # highest priority.
+        key = self._package_key(candidate_id)
```

Comment:
Ok, here's the problem. My understanding was incomplete--this does what I thought most of the time, but it can also cause a huge problem.

The policy came up with a candidate to propose. We hijack that here and suggest a different (equivalent) candidate. The next time around, the policy may try to suggest the original candidate again and we get stuck in an infinite loop. This would have been possible on `master`, too, but would not come up through our normal way of calling the solver because `prefer_installed` was set to `True`.